### ISSUE TITLE
 NoMethodError: undefined method `sub' for nil:NilClass when trying to delete record

### ISSUE THREAD

**Issue by saintberry** [chef-cookbooks/route53#76]
Hi, 

I'm successfully using this cookbook to create a record in a zone on Amazon Linux 2015.03 using OpsWorks. To get it working I had to specify fog_version to 1.29.0 as mentioned in another issue.

Create works fine, delete however throws an undefined method error that I can't pinpoint.

Initially I thought it could be to do with the version of fog being used. However I am able to jump onto the instance, fire up IRB after installing the correct version of fog and run the commands defined in the record.rb provider without issue:

```
> require 'fog'
> zone = Fog::DNS.new({ :provider => "AWS", :use_iam_profile => true }).zones.get("myzoneid")
> zone.records.get("sub.fqdn.", "A").destroy
> true
```

A stacktrace of the error follows:
```
[2015-06-26T05:25:28+00:00] INFO: No AWS credentials supplied, going to attempt to use IAM roles instead
 
================================================================================
Error executing action `delete` on resource 'route53_record[fe1.del.dev.nlm.io]'
================================================================================
 
 
NoMethodError
-------------
undefined method `sub' for nil:NilClass
 
 
Cookbook Trace:
---------------
/var/lib/aws/opsworks/cache.stage2/cookbooks/route53/providers/record.rb:57:in `zone'
/var/lib/aws/opsworks/cache.stage2/cookbooks/route53/providers/record.rb:69:in `record'
/var/lib/aws/opsworks/cache.stage2/cookbooks/route53/providers/record.rb:152:in `block in class_from_file'
 
 
Resource Declaration:
---------------------
# In /var/lib/aws/opsworks/cache.stage2/cookbooks/nlm_aws/recipes/route53_delete.rb
 
4: route53_record "delete route53 A record" do
5:     name  "#{node[:opsworks][:instance][:hostname]}.#{node[:nlm_aws][:route53][:fqdn]}"
6:     type  "A"
7:     action :delete
8: end
 
 
 
Compiled Resource:
------------------
# Declared in /var/lib/aws/opsworks/cache.stage2/cookbooks/nlm_aws/recipes/route53_delete.rb:4:in `from_file'
 
route53_record("fe1.del.dev.nlm.io") do
action [:delete]
retries 0
retry_delay 2
cookbook_name "nlm_aws"
recipe_name "route53_delete"
type "A"
end
 
 
 
[2015-06-26T05:25:28+00:00] INFO: Running queued delayed notifications before re-raising exception
[2015-06-26T05:25:28+00:00] ERROR: Running exception handlers
[2015-06-26T05:25:28+00:00] ERROR: Exception handlers complete
[2015-06-26T05:25:28+00:00] FATAL: Stacktrace dumped to /var/lib/aws/opsworks/cache.stage2/chef-stacktrace.out
[2015-06-26T05:25:28+00:00] ERROR: route53_record[fe1.del.dev.nlm.io] (nlm_aws::route53_delete line 4) had an error: NoMethodError: undefined method `sub' for nil:NilClass
[2015-06-26T05:25:28+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

To recap, I can successfully create records in a zone but get an error when trying to delete that record.

Edit: I've also tried fog v 1.31.0

**Comment by MhdAli** [comment]
Maybe because the record doesn't have root domain "." at the end? (I'm assuming here).

Another question when you create a record did you try a second time? check if it's trying to re-modify it or it says the record is up to date.

**Comment by saintberry** [comment]
Hi @MhdAli 

Same error with a "." at the end:

```
================================================================================
Error executing action `delete` on resource 'route53_record[fe1.del.dev.nlm.io.]'
================================================================================
 
 
NoMethodError
-------------
undefined method `sub' for nil:NilClass
 
 
Cookbook Trace:
---------------
/var/lib/aws/opsworks/cache.stage2/cookbooks/route53/providers/record.rb:57:in `zone'
/var/lib/aws/opsworks/cache.stage2/cookbooks/route53/providers/record.rb:69:in `record'
/var/lib/aws/opsworks/cache.stage2/cookbooks/route53/providers/record.rb:152:in `block in class_from_file'
 
 
Resource Declaration:
---------------------
# In /var/lib/aws/opsworks/cache.stage2/cookbooks/nlm_aws/recipes/route53_delete.rb
 
4: route53_record "delete route53 A record" do
5:     name  "#{node[:opsworks][:instance][:hostname]}.#{node[:nlm_aws][:route53][:fqdn]}."
6:     type  "A"
7:     action :delete
8: end
 
 
 
Compiled Resource:
------------------
# Declared in /var/lib/aws/opsworks/cache.stage2/cookbooks/nlm_aws/recipes/route53_delete.rb:4:in `from_file'
 
route53_record("fe1.del.dev.nlm.io.") do
action [:delete]
retries 0
retry_delay 2
cookbook_name "nlm_aws"
recipe_name "route53_delete"
type "A"
end
```

In regards to it checking multiple times... OpsWorks has these things called "lifecycle events". Basically you attach recipes to lifecycles (setup, configure, deploy, undeploy, shutdown). I have a recipe to create/update the record on setup and a recipe to delete the record on shutdown. When the shutdown lifecycle event runs the recipe used to create/update the record does not run. See the following chef info output from the *shutdown* event:

```
[2015-06-29T01:21:30+00:00] INFO: Processing chef_gem[nokogiri] action install (xml::ruby line 43)
[2015-06-29T01:21:30+00:00] INFO: Processing chef_gem[fog] action install (route53::default line 22)
[2015-06-29T01:21:30+00:00] INFO: Processing route53_record[fe1.del.dev.nlm.io.] action delete (nlm_aws::route53_delete line 4)
[2015-06-29T01:21:31+00:00] INFO: Getting record: fe1.del.dev.nlm.io. A
[2015-06-29T01:21:31+00:00] INFO: No AWS credentials supplied, going to attempt to use IAM roles instead

================================================================================
Error executing action `delete` on resource 'route53_record[fe1.del.dev.nlm.io.]'
================================================================================
```

**Comment by tas50** [comment]
Fog has been replaced with the aws-sdk so this is no longer a valid issue
